# Extracting latitude and longitude.
In this section i will extract the latitude, longitude and country code using Geonamecache. 
This is the second part of the project discovering the diseases in different cities. In  the previous section i extracted the cities and countries and observed that there are many missing values in the countries column. This is mainly the reason that the news data have more infromation about the diseases in the cities rather than the countries. I can still extract the countries just by using the city name, but in this notebook i will stick with the data for the cities and i will drop the countries. Also there are many cities that are in mulitple countries in that case we can just stick with the first choice that comes from the geonamecache module but only if there is no information given about the country. 

# Work Flow 
1.   Loading all the required modeules
2.   Loading the data that we have extracted in the last section.
3.   Extracting the longitude, latitude and country code using geonamecahe.
4.   Saving longitude, latitude and country code in the dataframe
5.   Droping the countries column because there are lots of values that are missing.
6.   Inspecting some extracted data to check all the values are correct.
 





# Importing all the libraries.

In [1]:
# Importing all the module
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
!pip install unidecode
import re
!pip install geonamescache
import geonamescache 
from nltk.tokenize import RegexpTokenizer
from unidecode import unidecode
from geonamescache import GeonamesCache


# Loading the dataset.

In [0]:
data = pd.read_json("headline_cities_and_countries.json")

In [0]:
#american = data['cities'] == "Can"
#data[american]

In [15]:

data.tail(33)

headline countries         cities
617                   Mad Cow Disease Detected in Lyon      None           Lyon
618    Memphis Encounters Severe Symptoms of Swine Flu      None        Memphis
619                 Zika Troubles come to Delray Beach      None   Delray Beach
620                How to Avoid Chikungunya in Hampton      None        Hampton
621                          Zika Reported in Sao Jose      None       Sao Jose
622            Case of Norovirus Reported in Stratford      None      Stratford
623             How to Avoid Hepatitis E in South Bend      None     South Bend
624                  Hepatitis C Outbreak in Bethlehem      None      Bethlehem
625                  Davenport is infested with Dengue      None      Davenport
626  Rumors about Swine Flu spreading in Suzhou hav...      None         Suzhou
627          Malaria Vaccine is now Required in Elyria      None         Elyria
628                    Zika case reported in Antioquia      None           None
629              Case of Swine Flu Reported in Tbilisi      None        Tbilisi
630               Chikungunya has not Left Pismo Beach      None           None
631                            Zika spreads to La Joya      None           None
632               Influenza Keeps Spreading in Suffolk      None        Suffolk
633  More people in Allentown are infected with Chl...      None      Allentown
634              Mad Cow Disease re-emerges in Bedford      None        Bedford
635  Authorities are Worried about the Spread of He...      None           Yiwu
636                        Zika Outbreak in Yurimaguas      None     Yurimaguas
637                          Dengue Outbreak in Easton      None         Easton
638                    Varicella has Arrived in Milton      None         Milton
639       New medicine wipes out Meningitis in Sevilla      None        Sevilla
640  Authorities are Worried about the Spread of Va...      None         Clovis
641          More Zika patients reported in Fort Worth      None     Fort Worth
642             Zika symptoms spotted in Boynton Beach      None  Boynton Beach
643                     Outbreak of Zika in Portoviejo      None     Portoviejo
644                       Influenza Exposure in Muscat      None         Muscat
645  Rumors about Rabies spreading in Jerusalem hav...      None      Jerusalem
646              More Zika patients reported in Indang      None         Indang
647  Suva authorities confirmed the spread of Rotav...      None           Suva
648         More Zika patients reported in Bella Vista      None    Bella Vista
649                     Zika Outbreak in Wichita Falls      None  Wichita Falls

In [16]:
data.shape

(650, 3)

In [17]:
data.isnull().sum()

headline       0
countries    635
cities        42
dtype: int64

In [0]:
gc = GeonamesCache()
cities = gc.get_cities()
num_cities = len(cities)

In [19]:
type(str(np.nan))

str

# Extracting latitude, longitude and country code.

In [0]:
#result =  gc.get_cities_by_name(data[['cities']])
#data['Latitude'] = np.nan
#data['Longitude'] = np.nan
#data['Country_code'] = str(np.nan)
def extracting_longitude_latitude(data):
  for index, row in data.iterrows():
    resulti = gc.get_cities_by_name(row['cities'])
    for  citiaa in resulti:
      city_info = list(citiaa.values())[0]
        #len(city_info)
      data.at[index, 'Latitude'] = city_info['latitude']
      data.at[index, 'Longitude'] = city_info['longitude']
      data.at[index, 'Country_code'] = city_info['countrycode']
  return data
     
#print(city_info)     

In [0]:
data = extracting_longitude_latitude(data)

In [22]:
data.head()

headline countries  ...  Longitude  Country_code
0                 Zika Outbreak Hits Miami      None  ...  -80.19366            US
1          Could Zika Reach New York City?      None  ...  -74.00597            US
2        First Case of Zika in Miami Beach      None  ...  -80.13005            US
3  Mystery Virus Spreads in Recife, Brazil    Brazil  ...  -34.88111            BR
4  Dallas man comes down with case of Zika      None  ... -123.31705            US

[5 rows x 6 columns]

In [0]:
#list(citiaa.values())[0]

In [0]:
#print(city_info)

In [25]:
data.tail(40)

headline  ... Country_code
610  Barstow Patient in Critical Condition after Co...  ...           US
611                      Hepatitis D Exposure in Aswan  ...           EG
612           Will Gonorrhea vaccine help East Moline?  ...           US
613       New medicine wipes out Influenza in Redlands  ...           US
614  How to Avoid Contaminated Meat when Visiting B...  ...           CH
615        More Zika patients reported in Kampong Speu  ...           KH
616           Alameda Residents Recieve Rabies vaccine  ...           US
617                   Mad Cow Disease Detected in Lyon  ...           FR
618    Memphis Encounters Severe Symptoms of Swine Flu  ...           US
619                 Zika Troubles come to Delray Beach  ...           US
620                How to Avoid Chikungunya in Hampton  ...           US
621                          Zika Reported in Sao Jose  ...          NaN
622            Case of Norovirus Reported in Stratford  ...           US
623             How to Avoid Hepatitis E in South Bend  ...           US
624                  Hepatitis C Outbreak in Bethlehem  ...           ZA
625                  Davenport is infested with Dengue  ...           US
626  Rumors about Swine Flu spreading in Suzhou hav...  ...           CN
627          Malaria Vaccine is now Required in Elyria  ...           US
628                    Zika case reported in Antioquia  ...          NaN
629              Case of Swine Flu Reported in Tbilisi  ...           GE
630               Chikungunya has not Left Pismo Beach  ...          NaN
631                            Zika spreads to La Joya  ...          NaN
632               Influenza Keeps Spreading in Suffolk  ...           US
633  More people in Allentown are infected with Chl...  ...           US
634              Mad Cow Disease re-emerges in Bedford  ...           US
635  Authorities are Worried about the Spread of He...  ...           CN
636                        Zika Outbreak in Yurimaguas  ...           PE
637                          Dengue Outbreak in Easton  ...           US
638                    Varicella has Arrived in Milton  ...           US
639       New medicine wipes out Meningitis in Sevilla  ...           ES
640  Authorities are Worried about the Spread of Va...  ...           US
641          More Zika patients reported in Fort Worth  ...           US
642             Zika symptoms spotted in Boynton Beach  ...           US
643                     Outbreak of Zika in Portoviejo  ...           EC
644                       Influenza Exposure in Muscat  ...           OM
645  Rumors about Rabies spreading in Jerusalem hav...  ...           IL
646              More Zika patients reported in Indang  ...           PH
647  Suva authorities confirmed the spread of Rotav...  ...           FJ
648         More Zika patients reported in Bella Vista  ...           US
649                     Zika Outbreak in Wichita Falls  ...           US

[40 rows x 6 columns]

# Testing 
Manually passing in some cities in geonamecache to check the longitude and latitude values.

In [26]:
test_Jerusalem = gc.get_cities_by_name('Jerusalem')
test_Jerusalem

[{'281184': {'admin1code': '06',
   'countrycode': 'IL',
   'geonameid': 281184,
   'latitude': 31.76904,
   'longitude': 35.21633,
   'name': 'Jerusalem',
   'population': 801000,
   'timezone': 'Asia/Jerusalem'}}]

In [27]:
test_Portoviejo = gc.get_cities_by_name('Portoviejo')
test_Portoviejo

[{'3652941': {'admin1code': '14',
   'countrycode': 'EC',
   'geonameid': 3652941,
   'latitude': -1.05458,
   'longitude': -80.45445,
   'name': 'Portoviejo',
   'population': 170326,
   'timezone': 'America/Guayaquil'}}]

In [28]:
data.isnull().sum()

headline          0
countries       635
cities           42
Latitude         67
Longitude        67
Country_code     67
dtype: int64

In [29]:
rslt_df = data.loc[data['Latitude'].isnull()] 
#rslt_df
#data =  data.drop(data.loc[data['cities'].isnull()])

rslt_df.shape
#data.drop(['cities'], axis=0)

(67, 6)

In [30]:
data.isnull().sum()

headline          0
countries       635
cities           42
Latitude         67
Longitude        67
Country_code     67
dtype: int64

# Handling Missing Values.
 As we can see that there are 635 values are missing in the countries column so we can drop that column because we are not getting any quality information from that column. Regarding the city column we will jsut keep that data that is not missiing in the cities column as well.

In [0]:
final_data  = data[data['cities'].notna()]

In [32]:
final_data.head()

headline countries  ...  Longitude  Country_code
0                 Zika Outbreak Hits Miami      None  ...  -80.19366            US
1          Could Zika Reach New York City?      None  ...  -74.00597            US
2        First Case of Zika in Miami Beach      None  ...  -80.13005            US
3  Mystery Virus Spreads in Recife, Brazil    Brazil  ...  -34.88111            BR
4  Dallas man comes down with case of Zika      None  ... -123.31705            US

[5 rows x 6 columns]

In [33]:
final_data.shape

(608, 6)

In [0]:
final_data = final_data.drop(['countries'],axis = 1)

In [35]:
final_data.head()

headline  ... Country_code
0                 Zika Outbreak Hits Miami  ...           US
1          Could Zika Reach New York City?  ...           US
2        First Case of Zika in Miami Beach  ...           US
3  Mystery Virus Spreads in Recife, Brazil  ...           BR
4  Dallas man comes down with case of Zika  ...           US

[5 rows x 5 columns]

In [36]:
final_data.isnull().sum()

headline         0
cities           0
Latitude        25
Longitude       25
Country_code    25
dtype: int64

We have 25 values missing in the longitude and latitude.
Let's examine why these values are missing.

In [37]:
missing_lat_long = final_data.loc[final_data['Latitude'].isnull()] 
missing_lat_long

headline  ... Country_code
7                Geneve Scientists Battle to Find Cure  ...          NaN
9                   Zika Infested Monkeys in Sao Paulo  ...          NaN
47                         18 new Zika Cases in Bogota  ...          NaN
64                    Second Zika Paitient in Brasilia  ...          NaN
82       Hepatitis B Vaccine is now Required in Quebec  ...          NaN
86                    Visitor to Cucuta contracts Zika  ...          NaN
101               Can Zika make it here to Vero Beach?  ...          NaN
115                  Spreading Zika ruining Cancun fun  ...          NaN
116             Cancun hit by Outbreak of Party Fever!  ...          NaN
167                  Arhus is infested with Bronchitis  ...          NaN
168                      Zika Troubles come to Niteroi  ...          NaN
199         More Zika patients reported in Joao Pessoa  ...          NaN
272    Schools in Yaounde Closed Due to Mumps Outbreak  ...          NaN
287            More Zika patients reported in Camacari  ...          NaN
341                     Zika symptoms spotted in Surat  ...          NaN
370  Rumors about Hepatitis D Spreading in Montreal...  ...          NaN
459      More Zika patients reported in Ribeirao Preto  ...          NaN
480            More Zika patients reported in Custodia  ...          NaN
515                    Zika spreads to San Luis Potosi  ...          NaN
522  Authorities are Worried about the Spread of Tu...  ...          NaN
524                New medicine wipes out HIV in Kihei  ...          NaN
539                     Zika case reported in Medellin  ...          NaN
549                      Zika Reported in Ciudad Acuna  ...          NaN
559            Kedougou tests new cure for Hepatitis C  ...          NaN
621                          Zika Reported in Sao Jose  ...          NaN

[25 rows x 5 columns]

# Data Quality.
 As we can see that we still have 25 missing values in the extracted columns (Longitude, Latitude, Country Code) . By carefully inspecting those city names it was found that those cities were not mentioned in the geonamecache. 
Data quality is a always problem in text related problems. We can still improve but we can get 100% accurate data. For this notebook we will just stick save the data and can see how we can pmrovise it in the next part 

# Saving the data For the next step in json Fromat.

In [0]:
import json

final_data.to_json (r'Longitude_latitude.json')

In [0]:
long_lat_data = pd.read_json('Longitude_latitude.json')

In [43]:
long_lat_data.head()

headline  ... Country_code
0                 Zika Outbreak Hits Miami  ...           US
1          Could Zika Reach New York City?  ...           US
2        First Case of Zika in Miami Beach  ...           US
3  Mystery Virus Spreads in Recife, Brazil  ...           BR
4  Dallas man comes down with case of Zika  ...           US

[5 rows x 5 columns]